netx/conn_test.go

```diff
@@ -0,0 +1,313 @@
+package netx
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/testutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+const testPayload = "test"
+
+func TestIntegrationWithConsulRoundRobinAndNetworkSend(t *testing.T) {
+
+	// start consul for discovery service
+	config, server := createTestConsulServer(t)
+	consulApiClient, err := api.NewClient(config)
+	require.NoError(t, err)
+	defer stopConsul(server)
+
+	// create listener acting as a service
+	service, err := tcpListener()
+	require.NoError(t, err)
+	defer service.Close()
+
+	// register service in consul
+	agent := consulApiClient.Agent()
+	err = agent.ServiceRegister(&api.AgentServiceRegistration{
+		ID:      "1",
+		Name:    "service-name",
+		Port:    service.Addr().(*net.TCPAddr).Port,
+		Address: service.Addr().(*net.TCPAddr).IP.String(),
+	})
+	require.NoError(t, err)
+
+	// create round robin writer writing by network and obtaining instances provided by consul
+	writer := RoundRobin(
+		DiscoveryServiceInstanceProvider(
+			"service-name",
+			time.Second,
+			NewConsulDiscoveryServiceClient(consulApiClient)),
+		NetworkSend)
+
+	// expect service read the payload
+	go func() {
+		conn, err := service.AcceptTCP()
+		require.NoError(t, err)
+
+		buf, err := ioutil.ReadAll(conn)
+		require.NoError(t, err)
+
+		assert.Equal(t, testPayload, string(buf))
+	}()
+
+	// write something
+	bytesSent, err := writer.Write([]byte(testPayload))
+
+	assert.NoError(t, err)
+	assert.Equal(t, len(testPayload), bytesSent)
+}
+
+func TestNetworkSendShouldReturnErrorWhenConnectionUnavailable(t *testing.T) {
+	listener, err := tcpListener()
+	require.NoError(t, err)
+	listener.Close()
+
+	bytesSent, err := NetworkSend(listener.Addr(), []byte("test"))
+
+	assert.Error(t, err)
+	assert.Zero(t, bytesSent)
+}
+
+func TestNetworkSendShouldReturnNumberOfSentBytes(t *testing.T) {
+	listener, err := tcpListener()
+	defer listener.Close()
+	require.NoError(t, err)
+
+	go func() {
+		conn, err := listener.AcceptTCP()
+		require.NoError(t, err)
+
+		buf, err := ioutil.ReadAll(conn)
+		require.NoError(t, err)
+
+		assert.Equal(t, testPayload, string(buf))
+	}()
+
+	bytesSent, err := NetworkSend(listener.Addr(), []byte(testPayload))
+
+	assert.NoError(t, err)
+	assert.Equal(t, len(testPayload), bytesSent)
+}
+
+func tcpListener() (*net.TCPListener, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+func TestDiscoveryServiceInstanceProviderShouldPeriodicallyUpdatesInstances(t *testing.T) {
```

Comment:
This does only check that everything is the same all the time - not what the name says. We should add another instance to see if it actually does work.